library/x509_create.c

```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+             */
+            case dn_after_tag:
             {
-                return( MBEDTLS_ERR_X509_ALLOC_FAILED );
+                /*
+                 * Skip whitespaces
+                 */
+                 while( cur < end && *cur == ' ' )
+                 {
+                    cur++;
+                 }
+
+                 if( *cur == '=' )
+                 {
+                     start = cur + 1;
+                     d = data;
+                     dn_cur_state = dn_before_value;
+                 }
+                 else
+                 {
+                     ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
+                     goto exit;
+                 }
             }
+            break;
+            /*
+             * Currently we have finished parsing the tag, an now starting to
+             * parse the dn name value
+             */
+            case dn_before_value:
+            {
+                value_len = 0;
+                /*
+                 * Skip whitespaces
+                 */
+                while( cur < end && *cur == ' ' )
+                {
+                    cur++;
+                    start++;
+                }
+
+                if( *cur == ',' )
+                {
+                    ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                    goto exit;
+                }
+                else if( *cur == '\"' )
+                {
+                    dn_cur_state = dn_in_quoted_value;
+                    start++;
+                }
+                else
+                {
+                    dn_cur_state = dn_in_unquoted_value;
+                    continue;
+                }
+            }
+            break;
 
-            // set tagType
-            cur->val.tag = attr_descr->default_tag;
-
-            while( c < end && *(c + 1) == ' ' )
-                c++;
-
-            s = c + 1;
-            in_tag = 1;
-        }
-
-        if( !in_tag && s != c + 1 )
-        {
-            *(d++) = *c;
+            /*
+             * Currently parsing a quoted value.
+             */
+            case dn_in_quoted_value:
+            {
+                if( *cur == '\"' && cur != end )
+                {
+
+                    if( value_len == 0 )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
+                        goto exit;
+                    }
+
+                    if( mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
+                                                       (unsigned char *) data,
+                                                       value_len ) == NULL )
+                    {
+                        ret = MBEDTLS_ERR_X509_ALLOC_FAILED;
+                        goto exit;
+                    }
+
+                    ( *head )->val.tag = attr_descr->default_tag;
+                    cur++;
+                    start = cur + 1;
+                    dn_cur_state = dn_after_quoted_value;
+                    continue;
```

Comment:
Cant we just leave it to break out of here normally and the cur will be incremented automatically?

Reply:
I don't think so. As you can see, we are skipping the quotation mark and modify `start` to point to `+2` from original `cur